tolstoy/src/tx_client.rs

```diff
@@ -62,52 +75,61 @@ pub struct TxClient {}
 
 impl TxReader for TxClient {
     fn all(sqlite: &rusqlite::Connection) -> Result<Vec<Tx>> {
-        // Make sure a=txInstant rows are first, so that we process
-        // all transactions before we process any transaction parts.
         let mut stmt = sqlite.prepare(
             "SELECT
                 e, a, v, tx, added, value_type_tag,
                 CASE a WHEN :txInstant THEN 1 ELSE 0 END is_transaction
-            FROM transactions ORDER BY is_transaction DESC, tx ASC"
+            FROM transactions"
         )?;
-        let rows: Vec<Result<RawTx>> = stmt.query_and_then_named(&[(":txInstant", &entids::DB_TX_INSTANT)], |row| -> Result<RawTx> {
-            Ok(RawTx {
+        let datoms: Vec<Result<RawDatom>> = stmt.query_and_then_named(&[(":txInstant", &entids::DB_TX_INSTANT)], |row| -> Result<RawDatom> {
+            Ok(RawDatom {
                 e: row.get(0),
                 a: row.get(1),
                 v: TypedValue::from_sql_value_pair(row.get(2), row.get(5))?,
                 tx: row.get(3),
-                added: row.get(4)
+                added: row.get(4),
+                is_transaction: row.get(6),
             })
         })?.collect();
 
         // It's convenient to have a consistently ordered set of results,
         // so we use a sorting map.
         let mut txes_by_tx = BTreeMap::new();
-        for row_result in rows {
-            let row = row_result?;
-            // Row represents a transaction.
-            if row.a == entids::DB_TX_INSTANT {
-                txes_by_tx.insert(row.tx, Tx {
-                    tx: row.tx,
-                    tx_instant: row.v,
-                    parts: Vec::new()
+        let mut tx_parts_by_tx = HashMap::new();
+
+        // On first pass, build our Txes and TxParts for each.
+        for datom_result in datoms {
```

Comment:
E.g., here, given that the sequence of datoms is going to be sorted by `tx` (especially if you make your query `ORDER BY tx, added`, which will use the index!), you can just running-collect by `datom.tx` and pick out the date as it passes by.